- Keyboard Shortcuts

JosephStar318 opened the pull request.

Commit:
- created graph operations and handled input from parley grap view

jonnydgreen commented on addons/parley/graph_operations/delete_operation.gd:
```diff
@@ -0,0 +1,31 @@
+# DeleteShortcut.gd
+extends GraphOperation
+class_name DeleteOperation
```
For all of these class_names, could we prefix with Parley to avoid clashes with other plugins? E.g. `ParleyDeleteGraphOperation`

jonnydgreen commented on addons/parley/views/graph_connection.gd:
```diff
@@ -0,0 +1,44 @@
+# Represents a connection between two GraphNodes in a GraphEdit
+
+extends Object
+class_name GraphConnection
```
Could we call this ParleyGraphEdge to keep consistent with other terminology in the codebase? (Same with the file name as well!)

jonnydgreen replied:
As long as we don't have breaking changes and the separation of concerns makes sense - I'm happy :)

JosephStar318 replied:
I decided not to do that because it was going to require a lot of code change

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -351,4 +351,172 @@ func set_selected_by_id(id: String, _goto: bool = true) -> void:
+var selectedConnections: Array[GraphConnection] = []
+var selectedNodes: Array[GraphNode] = []
+
+func _gui_input(event: InputEvent) -> void:
```
I think you can avoid a lot of this logic by using the signals built in to the GraphEdit Node. For example:
 - `delete_nodes_request`
 - `copy_nodes_request`
 - `cut_nodes_request`
 - `duplicate_nodes_request`

wdyt?

JosephStar318 replied:
i was not aware of those signals thanks for pointing that out. i will definitely use them!

jonnydgreen commented on addons/parley/views/graph_connection.gd:
```diff
@@ -0,0 +1,44 @@
+		return FAILED
+	return graph_edit.connect_node(from_node.name, from_port, to_node.name, to_port)
+
+func select() -> void:
```
I'll have a play to understand your comment on line 34 :)

JosephStar318 replied:
yeah its really weird. it seems the right ports indexes start from the index output_port_count

jonnydgreen replied:
> it seems the right ports indexes start from the index output_port_count

I had a play with this - which nodes are you selecting as I don't see the behaviour you are seeing (for dialogue nodes at least)?

With some custom logs for `all.ds`:
```
from_node DIALOGUE-27:<GraphNode#2755590705732>
output_port_count 1
from_port 0
Clicked connection:DIALOGUE-27:<GraphNode#2755590705732>:0 -> DIALOGUE-29:<GraphNode#2758140843496>:0
```

Are you selecting nodes with multiple output ports perhaps? (e.g. match nodes)

If it's of interest to you, a lot of this (or similar) functionality is available in `set_edge_colour` in `addons/parley/views/parley_graph_view.gd`. Ideally, if we're able to do things using the original ASTs (e.g. from the Edge AST), that would be preferable as everything is derived from this data - as always, happy to discuss! :)

JosephStar318 replied:
in `from_node.set_slot_color_right(output_port_count + from_port - 1, Color.DEEP_SKY_BLUE)` right side of the first node is from_node's right ports. If you log the `from_port` you get between 0 to (whatever port count - 1). But if you try to set the color of that index it changes the color of a different port. So i had an idea of maybe its starting from an offset so added the `output_port_count` into the equation and it worked. Thats where i was confused. As far as i can see you don't add extra offset and it works. I don't know why this requires that.

jonnydgreen replied:
interesting! What node were you seeing the `from_port ` being set to `1`?

JosephStar318 replied:
just to be clear, slot and node is the same in this context. and this was happening in every slot there was not any specific. only the output side though, input side works as expected as you can see in the code i used `to_port` as an index.

jonnydgreen replied:
> My initial thought was whether it's happening on some types of nodes

To illustrate this, take the dialogue node, this has the following slots:

<img width="161" height="127" alt="Screenshot 2025-10-03 at 10 53 57" src="https://github.com/user-attachments/assets/e0d503d4-140c-4e58-9e6a-59960724fe9a" />

Which matches up as expected and the `from_port` starts from `0` which corresponds to the slot count of `0`

However, with the condition node, things are a little different because of the layout:

<img width="191" height="159" alt="Screenshot 2025-10-03 at 10 51 07" src="https://github.com/user-attachments/assets/5a91a075-f7ee-4491-955d-88e56034b5d0" />

This doesn't match up as expected as the `from_port` starts can be `0` or `1` but the corresponding slots are actually: `1` and `2`. This is handled in the code in `condition_node.gd`:

<img width="619" height="551" alt="Screenshot 2025-10-03 at 10 53 23" src="https://github.com/user-attachments/assets/4118bca9-78af-41f7-a9a5-73a4018f37bd" />

Note, because the layout is not necessarily the same for each node in the graph view (match nodes will be even worse) I think we need to handle at the (graph) node level rather than generically (see above). I wonder if we can re-use the logic already available? (`select_from_slot`, `deselect_from_slot`)

wdyt?

JosephStar318 replied:
it was a derp on my part. I realized I didn't know the difference between node and slot. I will do some research on this hoping to find a good solution.

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -351,4 +351,172 @@ func set_selected_by_id(id: String, _goto: bool = true) -> void:
+		operation.do()
+		add_undo_operation(operation)
+
+func _save() -> void:
```
I think this one is better in the `main_panel.gd` as this is where the current save logic+button is - wdyt?

JosephStar318 replied:
i was thinking of moving most of the stuff that has importance to the graph to `main_panel.gd` using signals. Didn't want to do that until i get the pass from you :D

jonnydgreen replied:
Sounds good to emit events (e.g. node deleted) via signals as that keeps the existing separation of concerns consistent (e.g. AST manipulations in the `main_panel`) :) I guess this approach is highly relevant to this comment: https://github.com/bisterix-studio/parley/pull/30/files#r2361237850

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -351,4 +351,172 @@ func set_selected_by_id(id: String, _goto: bool = true) -> void:
+	print("Pasted copied node(s)")
+	# TODO: implement paste logic
+
+func add_undo_operation(operation: GraphOperation) -> void:
```
This undo/redo stack is really neat - nice one! 🚀

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -351,4 +351,172 @@ func set_selected_by_id(id: String, _goto: bool = true) -> void:
+				foundAnyConnection = true
+				var connection: GraphConnection = GraphConnection.new(from_node, from_port, to_node, to_port)
+				connection.select()
+				selectedConnections.append(connection as GraphConnection)
+				onUnselect.append(func() -> void:
+					connection.unselect()
+					selectedConnections.erase(connection)
+				)
```
What's the reason for capturing the selected connections here?

JosephStar318 replied:
In delete operation i wanted to delete all selected connections

jonnydgreen commented on addons/parley/graph_operations/graph_operation.gd:
```diff
@@ -0,0 +1,12 @@
+	# override in child class
+	push_error("undo() not implemented in subclass")
+
+func do() -> void:
```
should we call this `redo`?

jonnydgreen reviewed (commented):
Thanks for submitting! Great work so far! I've done a first pass and left some comments :)

One general comment is that upon an execution of an operation, we need to update the underlying AST and clear up any open editors in the dock. As you mentioned on the linked issue, this means we need to notify the main_panel that something has happened. The implementation for this probably depends on the outcome of some of my comments - happy to help when these are sorted/agreed :)

Some thoughts for the future:
 - Testing this will be interesting - if it's not easily possible, no problem at all but it will be interesting to see what we can do with GUT!
 - We'll need to think about adding some documentation as well. Nothing major, just something to say what keyboard commands are available etc. Again, happy to show you how the docs work when the time comes :)

Any questions, give me a shout!

JosephStar318 commented:
so here is a question, after I managed to delete a node successfully and save the dialogue ast. But undoing it will require for me to re-add the node to the dialogue ast and I couldn't find a capable method for doing that. only method i found is
`add_ast_node` which expects a dictionary. I wanted to be able to update the give the node of type ParleyGraphNode or maybe its id, and change the dialogue ast. How can I do that?

Commit:
- selecting multibple connections, deleting connection and nodes works, undo delete WIP

Commit:
- naming changes

jonnydgreen commented:
> No worries, i couldn't find time to continue my work on this anyways. Maybe this weekend i will get some stuff done

Thanks and absolutely no rush on your end as well :)

jonnydgreen commented:
> I wanted to be able to give the node of type ParleyGraphNode or maybe its id, and change the dialogue ast. How can I do that?

Good question! `add_ast_node` is probably not what we want as this is for use by the importer - I'll make a note to document this as I'll admit, it isn't clear :) How are you currently storing deleted nodes? I think we may want to store the AST (e.g. `ParleyMatchNodeAst`) in the history!

If storing as an AST, I would recommend defining a new method that accepts `NodeAst` as an input, given that different nodes store different data. This would then add it to the `DialogueSequenceAst`. So something like:

```gdscript
func add_node_from_ast(node_ast: ParleyNodeAst, insertion_index: int) -> ParleyNodeAst:
  # Probably need to make sure the provided ID doesn't already exist
  # Also worth storing the insertion index somewhere as well?
  nodes.insert(insertion_index, node_ast)
  _emit_dialogue_updated()
  return node_ast
```

wdyt?

JosephStar318 commented:
> > I wanted to be able to give the node of type ParleyGraphNode or maybe its id, and change the dialogue ast. How can I do that?
> 
> Good question! `add_ast_node` is probably not what we want as this is for use by the importer - I'll make a note to document this as I'll admit, it isn't clear :) How are you currently storing deleted nodes? I think we may want to store the AST (e.g. `ParleyMatchNodeAst`) in the history!
> 
> If storing as an AST, I would recommend defining a new method that accepts `NodeAst` as an input, given that different nodes store different data. This would then add it to the `DialogueSequenceAst`. So something like:
> 
> ```gdscript
> func add_node_from_ast(node_ast: ParleyNodeAst, insertion_index: int) -> ParleyNodeAst:
>   # Probably need to make sure the provided ID doesn't already exist
>   # Also worth storing the insertion index somewhere as well?
>   nodes.insert(insertion_index, node_ast)
>   _emit_dialogue_updated()
>   return node_ast
> ```
> 
> wdyt?

i was going to store the nodes itself without queuefree them. but storing asts makes much more sense. Only problem is that i don't have access to the ast from the parleygraphnode itself. if that was a variable inside maybe i could do that. or let me know if there is a way to get ast from graphnode.

JosephStar318 commented:
Okay i will use the id to get the ast. But to make things loosely coupled I prefer handling the operations inside the operation. If we need some components or scripts we can insert them when creating the operation. 
The current undo redo history is stored in the main_panel.gd. so my question here would be does main_panel reset when you load a new dialog? If it resets we don't need to do anything it will just clear the history as well. but if it doesn't change like you said we need to group by dialog sequence. it should not be a big deal, just going to make a dictionary of history.

jonnydgreen commented:
> things loosely coupled I prefer handling the operations inside the operation

Agreed - nice one!

> so my question here would be does main_panel reset when you load a new dialog

I believe it would as it should call the `refresh` method so as you mention, it should be okay - one for us to double check! :)

JosephStar318 commented:
> ```gdscript
> func add_node_from_ast(node_ast: ParleyNodeAst, insertion_index: int) -> ParleyNodeAst:
>   # Probably need to make sure the provided ID doesn't already exist
>   # Also worth storing the insertion index somewhere as well?
>   nodes.insert(insertion_index, node_ast)
>   _emit_dialogue_updated()
>   return node_ast
> ```
Does insertion_index matter? Couldn't we just append the node_ast? As i understand, the nodes array in dialogue_sequence_ast.gd are handled with like push_back functions which doesn't care the index.

Commit:
- delete redo undo cycle mostly done

Commit:
- delete operation improvements

- undo redo history grouped by dialoge sequence ast

JosephStar318 commented:
I have finally finished the delete operation and undo redo functionalities are working smoothly. even between two different dialogue sequence. Needed to add some get or find methods to get access to ast's. Let me know your feedbacks

jonnydgreen commented:
> I have finally finished the delete operation and undo redo functionalities are working smoothly. even between two different dialogue sequence. Needed to add some get or find methods to get access to ast's. Let me know your feedbacks

Awesome stuff! I will have a look asap

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -351,4 +356,135 @@ func set_selected_by_id(id: String, _goto: bool = true) -> void:
+			#Ctrl + Y
+			elif key_event.ctrl_pressed and key_event.keycode == KEY_Y:
+				redo_request.emit()
+			#Ctrl + S
```
On MacOS, I think this is Cmd + S, if you're able to account for this as well, that would be fab!

jonnydgreen replied:
yeah, it differs by replacing ctrl with cmd in a few common commands, e.g. these are the keybindings for godot on MacOS:

<img width="629" height="139" alt="Screenshot 2025-11-02 at 16 27 37" src="https://github.com/user-attachments/assets/9b089719-9ff8-4f05-a4ba-43c6aefd3d27" />

I wonder if we can do something smart from reading the value from `ui_*` and then using that?

JosephStar318 replied:
that totally makes sense!

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -351,4 +356,135 @@ func set_selected_by_id(id: String, _goto: bool = true) -> void:
+			elif key_event.keycode == KEY_DELETE:
+				delete_selected.emit()
+		
+func _handle_mouse_select(mouse_event: InputEventMouseButton) -> void:
```
Just looking at this function logic, it seems a little quadratic. I wonder how this will behave for large dialogue sequences - especially since it's running on every click event, which is a very common operation. I suspect we may have performance issues with this, so I wonder if this is a feature we could maybe turn off in project settings and allow users to opt-in! wdyt?

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -351,4 +356,135 @@ func set_selected_by_id(id: String, _goto: bool = true) -> void:
+			# print("From: ",from_node ," To", to_node," Comparing x: ", from_pos.x, " ", pointer_pos.x, " ", to_pos.x , " distance: ", distance)
+
+			if  distance <= CLICK_DISTANCE:
+				# basically ctrl+click selects multiple
```
```suggestion
				# basically ctrl/cmd+click selects multiple
```

jonnydgreen commented on addons/parley/views/parley_graph_edge.gd:
```diff
@@ -0,0 +1,63 @@
+var previousToColor: Color
+var selected: bool
+
+func _init(_edge_ast: ParleyEdgeAst, _from_node: ParleyGraphNode, _from_port: int, _to_node: ParleyGraphNode, _to_port: int) -> void:
```
Nit: in `_init` methods, I think it's generally good practice to add a default value to the parameters so you can call it as follows: `ParleyGraphEdge.new()` - I can't recall where I read this, but I think I read it in the gdscript tutorial

jonnydgreen replied:
See the docs comment here: https://docs.godotengine.org/en/stable/classes/class_object.html#class-object-private-method-init

<img width="826" height="231" alt="Screenshot 2025-11-24 at 21 04 10" src="https://github.com/user-attachments/assets/e73f660a-dd0d-4ad5-904f-f73e9125b7ca" />

So if we were to call `duplicate` on this class, it would fail. I generally prefer to use factory methods for this reason.

However, if we're only ever instantiating this directly then all good! But maybe worth adding a comment to illustrate this just in case :)

JosephStar318 replied:
Alright I'll add a comment

jonnydgreen commented on addons/parley/views/parley_graph_edge.gd:
```diff
@@ -0,0 +1,63 @@
+
+func connect_node(graph_view: ParleyGraphView) -> int:
+	if not graph_view:
+		return FAILED
```
I've been searching for something like this for ages (I've previously used the specific: `ERR_` codes) - TIL!

jonnydgreen reviewed (commented):
Hi! Really good progress! :) Just reviewed the code now and left some comments - let me know what you think!

I also had a play with your changes just now as well and I couldn't get the redo/undo to work (delete is fine) - I got the following errors in the console while playing around with: `all.ds` - have you come across these errors?

<img width="1069" height="145" alt="Screenshot 2025-10-26 at 12 47 47" src="https://github.com/user-attachments/assets/ddc7441f-ec47-46c5-9447-a4ebcb0fe258" />
<img width="1058" height="48" alt="Screenshot 2025-10-26 at 12 47 02" src="https://github.com/user-attachments/assets/4e1ff3a5-d663-4f15-b16f-294675bd9fbd" />

I also noticed a slight bit of colour weirdness when creating a new edge, every out going connection turns blue instead of staying light green. I don't know if you've noticed this?

<img width="220" height="286" alt="Screenshot 2025-10-26 at 12 48 17" src="https://github.com/user-attachments/assets/da26cba8-a6fa-49c9-8cdc-ec9980695b2d" />

Just FYI, I'm starting a new job next week so I may be slow to respond over the next week but I will have some time on weekends :)

JosephStar318 commented:
> Hi! Really good progress! :) Just reviewed the code now and left some comments - let me know what you think!
> 
> I also had a play with your changes just now as well and I couldn't get the redo/undo to work (delete is fine) - I got the following errors in the console while playing around with: `all.ds` - have you come across these errors?
> 
> <img alt="Screenshot 2025-10-26 at 12 47 47" width="1069" height="145" src="https://private-user-images.githubusercontent.com/31290910/505722833-ddc7441f-ec47-46c5-9447-a4ebcb0fe258.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjE0OTY5MjksIm5iZiI6MTc2MTQ5NjYyOSwicGF0aCI6Ii8zMTI5MDkxMC81MDU3MjI4MzMtZGRjNzQ0MWYtZWM0Ny00NmM1LTk0NDctYTRlYmNiMGZlMjU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDI2VDE2MzcwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNiNjNkZGQ4ZmJhOTdlYTU4MzFlYmY3MDI2ZjhiYjk1YzI5ZTJjYzQ0MzQ4NTJjZDkxMmNkNjM1YmIxZDllZDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.F3i75pHjVTKXR6mmlJ4Z3OSmm_RfvON9lCXody7TuiI"> <img alt="Screenshot 2025-10-26 at 12 47 02" width="1058" height="48" src="https://private-user-images.githubusercontent.com/31290910/505722835-4e1ff3a5-d663-4f15-b16f-294675bd9fbd.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjE0OTY5MjksIm5iZiI6MTc2MTQ5NjYyOSwicGF0aCI6Ii8zMTI5MDkxMC81MDU3MjI4MzUtNGUxZmYzYTUtZDY2My00ZjE1LWIxNmYtMjk0Njc1YmQ5ZmJkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDI2VDE2MzcwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQxMGQwMWE1YjQyYmEwOTNmMGMwZjEyNzgyNGRlMzEzNWUzYjJiMTA5ZThiYmZlYWNkZGRlMmFiYjRmZWEyMzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.9Vxb3IUaFu7OFqFN_HHRHkJP-3b7b-9rrzg65WT9iGk">
> I also noticed a slight bit of colour weirdness when creating a new edge, every out going connection turns blue instead of staying light green. I don't know if you've noticed this?
> 
> <img alt="Screenshot 2025-10-26 at 12 48 17" width="220" height="286" src="https://private-user-images.githubusercontent.com/31290910/505723318-da26cba8-a6fa-49c9-8cdc-ec9980695b2d.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjE0OTY5MjksIm5iZiI6MTc2MTQ5NjYyOSwicGF0aCI6Ii8zMTI5MDkxMC81MDU3MjMzMTgtZGEyNmNiYTgtYTZmYS00OWM5LThjZGMtZWM5OTgwNjk1YjJkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDI2VDE2MzcwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVmNTU0MDg4YjhmOWVjNmE0Mzc5NzJkMzExMTc0M2Q5MjNiNDcwMzBiYjUzOTYzNjMxNDMzMmJmY2JiMWE4M2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.HGJwYO3b3r5QuXnHNGjUfEaClO48HHS41yzEsQNqWkY">
> Just FYI, I'm starting a new job next week so I may be slow to respond over the next week but I will have some time on weekends :)

i just realized you could also set the color of ports. i think i changed the slot color and that is why it affects every edge.

I will take look at the errors probably going to need a lot more testing.

The  undo redo didn't work assuming you were on a mac and based on the other comments you wrote there is no support for it right?

Commit:
- bug fixes and improvements

JosephStar318 commented:
sorry for the delay i was super busy these couple of weeks.

as for the progress, i implemented the macos support and learned that port color doesn't exist. it's a default behavior by godot to change all the connections one side color when you set the slot color. there might be an advanced solution which i am not aware of. Would like to see if you encounter any more errors after this last commit.

JosephStar318 committed:
changed a comment

Co-authored-by: Jonny Green <me+github@jonnydgreen.com>

Commit:
- inserted new lines

Commit:
Merge branch 'keyboard-shortcuts' of https://github.com/JosephStar318/parley into keyboard-shortcuts

Commit:
- print rich and removed unnecessary prints

Commit:
- edge selection color improvements

- storing node_ids instead of nodes to prevent freed instance errors

jonnydgreen commented:
Just getting back up to speed with everything now :)

jonnydgreen commented:
Hello! I've managed to smooth out all the issues I was seeing (minus one that I can no longer replicate :))

Firstly, I can now see redo working perfectly! Super nice work! 🚀 The reason why I hadn't seen it before is for some reason my Godot editor key map was interpreting Cmd+shift+z as undo rather than redo 🤷  However, after using an alternative key map - it worked great!

I've tested on large dialogue sequences and all looking good on there as well!

Thank you for implementing the colour improvements! I've just reviewed and I found a couple of issues unfortunately, however the good news is I've also found the solutions to resolve them which should also reuse existing logic for selecting and deselecting edges :) lemme know what you think!

**Issue 1**: when un/selecting the Condition Node false edge (red), it still used green/blue rather than red/blue

The below changes should resolve issue 1. There is some existing logic (that is used by the edge editor) to use the correct colour depending on the node. From my testing today, it now works pretty reliably so I'd recommend implementing the below if at all possible! wdyt? (apologies for the inconsistent naming of deselect/unselect btw, this is some tech-debt I've not got round to yet) 

<img width="1222" height="551" alt="Screenshot 2026-02-01 at 21 29 56" src="https://github.com/user-attachments/assets/19442d39-ae41-4b60-a70a-bb52cb413ea6" />

**Issue 2**: when creating a new edge. It marks it as immediately selected.

This is because the mouse release event fires upon edge creation and the cursor is still within the edge area. The first question about this issue is, whether it's an issue at all. I.e. do we actually want to select an edge upon edge creation? Unfortunately, I've seeing a lot of issues with the incorrect edge being selected when multiple edges are connected to a node and we end up selecting the wrong edge. As a result, my vote is we should have it unselected and retain the default behaviour of Parley. We would therefore need to make the follow change (note, we'd need to add a signal for `connection_requested`) - wdyt?

<img width="874" height="240" alt="Screenshot 2026-02-01 at 21 29 01" src="https://github.com/user-attachments/assets/96f75d8d-63f5-46b9-a18f-ae66e6c17c95" />

Other than that, everything is looking awesome so once the above is resolved, very happy from my side to look at undo/redo for node creation :)

JosephStar318 commented:
i am good with both of your solutions. I will send a change when i get the change. also the edge selection is currently not accurate i am aware but once i implement that bezier approximation it should fix that issue hopefully!

jonnydgreen commented:
> I will send a change when i get the change. also the edge selection is currently not accurate i am aware but once i implement that bezier approximation it should fix that issue hopefully!

Fantastic! Thanks and thanks! :)

Commit:
- color selection bug fix

- implemented bezier approximation

Commit:
- added create node opertion and connection to empty node operation

Commit:
- added duplicate operation

JosephStar318 commented:
I have done your suggested changes. also i have committed and improved edge selection with bezier calculation. And i didn't imagine it would work this smoothly, you will see the result yourself :D
Then i added node creation undo redo support as well as duplicate with cmd/ctrl +d. that is also supported with undo redo. 
Probably tomorrow i will take a look at the copy and paste. It should be easy but need to offset the nodes to the cursor position somehow. Its getting real close to its final shape though.

Commit:
- bug fixes and improvements

Commit:
- group selection fix and removed logs

JosephStar318 commented:
I ended up fixing a lot of bugs today. So could not hookup copy paste stuff. Also I couldn't find where you set the name of the node.
<img width="597" height="671" alt="image" src="https://github.com/user-attachments/assets/d14d926a-635c-421d-bf1c-671455861155" />
duplicating group node and setting the node name directly didn't change the name

jonnydgreen commented:
cool! Taking a look now! :)

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -29,17 +34,21 @@ func _exit_tree() -> void:
-	clear_connections()
+	# _unselect_connections()
+	# _clear_selected_nodes()
+	# clear_connections()
```
I guess these will be uncommented at some point?

JosephStar318 replied:
i will remove them actually. everytime a node gets selected or deselected clear was running thus causing the selection to be cleared when it should not. instead i made selections safe to reuse after refresh

jonnydgreen commented on addons/parley/graph_operations/parley_duplicate_operation.gd:
```diff
@@ -0,0 +1,58 @@
+class_name ParleyDuplicateOperation
```
For all of the added files in this PR, could you add a `@tool` to the top of each - thanks! :)

JosephStar318 replied:
will do!

jonnydgreen commented on addons/parley/graph_operations/parley_duplicate_operation.gd:
```diff
@@ -0,0 +1,58 @@
+	var node_names: Array[String]
+
+	for node_id: String in selected_node_ids:
+		var node : ParleyGraphNode = graph_view.find_node_by_id(node_id)
```
As this is always the source of truth, I would highly recommend finding by ID in the AST instead. This is because not all data is included in graph nodes, whereas all the data can be found in AST nodes.

This will have additional benefits in the future as well! :)

JosephStar318 replied:
can do. but i think this section will be changed with copy methods for each node ast?

jonnydgreen commented on addons/parley/graph_operations/parley_node_data.gd:
```diff
@@ -0,0 +1,10 @@
+class_name NodeData
```
I believe this is made redundant by the suggestion below!

JosephStar318 replied:
what was the suggestion?

jonnydgreen replied:
ah sorry I should've been specific, I was referring to the the `.copy()` suggestion

jonnydgreen commented on addons/parley/graph_operations/parley_duplicate_operation.gd:
```diff
@@ -0,0 +1,58 @@
+		var ast_node: ParleyNodeAst = graph_view.ast.add_new_node(node.type, node.position_offset + Vector2.DOWN * 200 + Vector2.RIGHT * 200 )
+		if ast_node:
+			node_id = ast_node.id
+			var connections : Array[ParleyGraphEdge] = ParleyGraphUtils.get_connections_for_node(graph_view, node_id)
```
I'm not sure we should be doing duplicating edges, I vote we keep this simple and just duplicate the nodes only!

jonnydgreen commented on addons/parley/graph_operations/parley_duplicate_operation.gd:
```diff
@@ -0,0 +1,58 @@
+
+	for node_id: String in selected_node_ids:
+		var node : ParleyGraphNode = graph_view.find_node_by_id(node_id)
+		var ast_node: ParleyNodeAst = graph_view.ast.add_new_node(node.type, node.position_offset + Vector2.DOWN * 200 + Vector2.RIGHT * 200 )
```
To keep with existing patterns, provide extra flexibility, simplicity, and save on adding node specific logic to these operations, could we instead add a method to each ParleyNodeAst type called `.copy()`? (I don't trust the builtin `.duplicate` haha and plus, we have more control) This would copy all the relevant fields of each node and then add a new function called `add_node_ast` to the ast:


```suggestion
		var ast_node = graph_view.ast.find_node_by_id(node_id)
		var new_ast_node = ast_node.copy()
		new_ast_node.position = new_ast_node.position + Vector2.RIGHT * 200
		graph_view.ast.add_node_ast(new_ast_node) # this would create a new ID internally
```

This should hopefully be the solution to your group node question as well! :)

JosephStar318 replied:
you are right about moving this to ast other than node. it might break when doing undo and redo operations and if it could not find the node it would cause problems. my only problem with this node_ast doesn't have a name property.  or is it just the resource name we need to set the node's name

JosephStar318 replied:
also does the ids of GraphNode's and NodeAst's not match?

jonnydgreen commented on addons/parley/main_panel.gd:
```diff
@@ -295,9 +297,10 @@ func _on_insert_id_pressed(type: ParleyDialogueSequenceAst.Type) -> void:
 		return
-	var ast_node: Variant = dialogue_ast.add_new_node(type, (graph_view.scroll_offset + graph_view.size * 0.5) / graph_view.zoom)
-	if ast_node:
-		await refresh()
```
Is this removal of `refresh` intentional?

JosephStar318 replied:
yes it was causing problems with selecting nodes inside group nodes i think

jonnydgreen replied:
Okay thanks, I will have a play as this was needed in the original release to ensure that we handle things deterministically :)

I've been having a think last week and given the subtlety of needing this here in the first place and the size of this PR, I think it would be a good to idea to have this feature behind feature flag for this PR as an opt-in! This should be easy enough with an addition to `addons/parley/settings.gd` - let me know if you need a hand with this :)

jonnydgreen commented on addons/parley/main_panel.gd:
```diff
@@ -560,14 +563,15 @@ func _on_graph_view_connection_to_empty(from_node_name: StringName, from_slot: i
-	var ast_node_variant: Variant = dialogue_ast.add_new_node(ParleyDialogueSequenceAst.Type.DIALOGUE, ((graph_view.scroll_offset + release_position) / graph_view.zoom) + Vector2(0, -90))
-	if ast_node_variant and ast_node_variant is ParleyNodeAst:
-		var ast_node: ParleyNodeAst = ast_node_variant
-		await refresh()
```
Is this removal of `refresh` intentional?

JosephStar318 replied:
i think this is the same reason with above comment

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -351,4 +360,295 @@ func set_selected_by_id(id: String, _goto: bool = true) -> void:
+	)
+
+
+func get_distance_to_bezier(point: Vector2, p0: Vector2, p1: Vector2, p2: Vector2, p3: Vector2) -> float:
```
🔥

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -351,4 +360,295 @@ func set_selected_by_id(id: String, _goto: bool = true) -> void:
+	for con: ParleyGraphEdge in selected_connections:
+		if con.from_node == _from_node && con.from_port == _from_port && con.to_node == _to_node && con.to_port == _to_port:
+			return con
+	return null
```
Just to avoid future confusion, could we align the terminology to `edge` rather than connection - thanks! :)

JosephStar318 replied:
will do!

jonnydgreen commented on addons/parley/graph_operations/parley_duplicate_operation.gd:
```diff
@@ -0,0 +1,58 @@
+		var ast_node: ParleyNodeAst = graph_view.ast.add_new_node(node.type, node.position_offset + Vector2.DOWN * 200 + Vector2.RIGHT * 200 )
+		if ast_node:
+			node_id = ast_node.id
+			var connections : Array[ParleyGraphEdge] = ParleyGraphUtils.get_connections_for_node(graph_view, node_id)
```
I'm not sure we should be duplicating edges as we would likely introduce expected behaviour and complexity with this. I vote we keep this simple and just duplicate the nodes only! :)

JosephStar318 replied:
yeah my plan was to do save those for later. i haven't hooked that logic so this was just a ground work for that

jonnydgreen reviewed (commented):
Great stuff and outstanding work with the bezier curve! As you say, we are getting there! 🚀 (And of course, no rush on the copy/paste work!)

I haven't finished my review yet as I've mainly focussed on the duplication operator this evening - I'll continue this over the weekend as I'm busy every other evening this week unfortunately!

In the meantime, I had a play and it's working nicely at first glance:) I did notice the following issues, I don't know if you've come across them?

1. Duplication nodes don't copy values
2. Weirdness happens with the edges when I do the following on `all.ds`:
	a. Duplicate node 27
    b. Add edge from 27 -> 29
    c. Undo change
    d. ==> Weirdness with edges
3. Duplicating with multiple selections didn't work for me

Commit:
- added @tool

- changed name connection to edge
- removed some of the commented code
- added copy function to nodeast
- improved duplicate operation
- added copy paste support

JosephStar318 commented:
done your suggested changes and added copy paste support! please take a look

jonnydgreen commented:
> done your suggested changes and added copy paste support! please take a look

Fantastic! :) I will take a look this week and answer all your other Q's!

jonnydgreen commented on addons/parley/models/node_ast.gd:
```diff
@@ -36,10 +36,10 @@ func _init(_id: String = "", _position: Vector2 = Vector2.ZERO) -> void:
+		id = new_id
+	elif not id or id == id_prefix or not id.begins_with(id_prefix):
 		id = new_id if new_id.begins_with(id_prefix) else "%s%s" % [id_prefix, new_id]
-#endregion
```
I think this may have been accidental, could we revert this please - thanks :)

JosephStar318 replied:
<img width="550" height="134" alt="image" src="https://github.com/user-attachments/assets/2efed617-d004-45bb-baea-c432ce9b5787" />

this was needed because i couldn't override an id to a newly created instance of node_ast via .duplicate call. If you have a better workaround i will gladly accept it.

JosephStar318 replied:
copy function is not needed anymore and removed but setting the id externally still requires this change, i believe on of your comments you made a suggestion of changing comment on the id setter so i take it this is resolved

jonnydgreen replied:
ah sorry - I was referring to the deletion of the `#endregion` only - the other changes are all good :)

jonnydgreen commented on addons/parley/models/dialogue_sequence_ast.gd:
```diff
@@ -144,6 +144,24 @@ func add_new_node(type: Type, position: Vector2 = Vector2.ZERO) -> ParleyNodeAst
+	# Probably need to make sure the provided ID doesn't already exist
+	for node : ParleyNodeAst in nodes:
+		if node.id == node_ast.id:
+			print("Id already exists in dialogue sequence")
```
Just to keep errors and warnings consistent - thanks!

```suggestion
			push_warning(ParleyUtils.log.warn_msg("ID already exists in Dialogue Sequence. Not adding to AST."))
```

jonnydgreen commented on addons/parley/models/dialogue_sequence_ast.gd:
```diff
@@ -144,6 +144,24 @@ func add_new_node(type: Type, position: Vector2 = Vector2.ZERO) -> ParleyNodeAst
 	return ast_node
 
+func copy_node_ast(ast_node: ParleyNodeAst) -> ParleyNodeAst:
+	print_rich(ParleyUtils.log.info_msg('Inserting new Node into the AST of type: %s' % [ast_node.type]))
```
```suggestion
	print_rich(ParleyUtils.log.info_msg('Copying Node with ID %s and Type %s into the AST' % [ast_node.id, ast_node.type]))
```

jonnydgreen commented on addons/parley/models/dialogue_sequence_ast.gd:
```diff
@@ -144,6 +144,24 @@ func add_new_node(type: Type, position: Vector2 = Vector2.ZERO) -> ParleyNodeAst
+			print("Id already exists in dialogue sequence")
+			return
+			
+	# Also worth storing the insertion index somewhere as well?
```
So we can track later :)

```suggestion
	# TODO: also worth storing the insertion index somewhere as well
```

jonnydgreen commented on addons/parley/models/node_ast.gd:
```diff
@@ -36,10 +36,10 @@ func _init(_id: String = "", _position: Vector2 = Vector2.ZERO) -> void:
-	if not id or id == id_prefix or not id.begins_with(id_prefix):
+	if id.begins_with(id_prefix):
+		id = new_id
+	elif not id or id == id_prefix or not id.begins_with(id_prefix):
```
I assume this is needed because of the copying functionality?

If so, could we change to the following as the comment no longer applies:

```suggestion
    # Only allow change when copying or if not defined
	if new_id.begins_with(id_prefix):
		id = new_id
	elif not id or id == id_prefix or not id.begins_with(id_prefix):
```

JosephStar318 replied:
ofcourse!

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -29,17 +34,18 @@ func _exit_tree() -> void:
 	if arrange:
 		arrange_nodes()
+	
+	call_deferred("refresh_edges")
```
Is this addition needed? It's unfortunately conflicting with some of the existing refreshing functionality (e.g. unexpected nodes get reselected after clicking the refresh button) so I'd be tempted to remove it if at all possible! Or we can put it behind the feature flag :)

JosephStar318 replied:
it was to restore the states of selected edges after undoing. if you could tell me your scenario i can make a fix i think

jonnydgreen replied:
I can't remember the scenario unfortunately! To be on the safe side, could we add this behind a feature flag? Thanks! :)

```suggestion
	if ParleySettings.get_setting(Constants.EDITOR_IS_KEYBOARD_SHORTCUTS_ACTIVE, false): 
		call_deferred("refresh_edges")
```

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -351,4 +357,281 @@ func set_selected_by_id(id: String, _goto: bool = true) -> void:
+var _node_selection_triggered: bool
+var moving_nodes: bool
+
+func _on_node_selected(node: Node) -> void:
```
Let's return early with the Feature Flag here

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -351,4 +357,281 @@ func set_selected_by_id(id: String, _goto: bool = true) -> void:
+	# print("selected callback ", node_id, " ", selected_node_ids.size())
+
+
+func _on_node_deselected(node: Node) -> void:
```
Let's return early with the Feature Flag here

JosephStar318 replied:
you mean in every callback function?

JosephStar318 replied:
done!

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -351,4 +357,281 @@ func set_selected_by_id(id: String, _goto: bool = true) -> void:
+	selected_node_ids.erase((node as ParleyGraphNode).id)
+
+
+func _on_edges_deselected() -> void:
```
Let's return early with the Feature Flag here

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -351,4 +357,281 @@ func set_selected_by_id(id: String, _goto: bool = true) -> void:
+		callable.call()
+
+
+func _gui_input(event: InputEvent) -> void:
```
Let's return early with the Feature Flag here

jonnydgreen commented on addons/parley/main_panel.gd:
```diff
@@ -295,9 +297,10 @@ func _on_insert_id_pressed(type: ParleyDialogueSequenceAst.Type) -> void:
+	var node_position: Vector2 = (graph_view.scroll_offset + graph_view.size * 0.5) / graph_view.zoom
+	var create_node_operation: ParleyCreateNodeOperation = ParleyCreateNodeOperation.new(graph_view, type, node_position)
+	create_node_operation.do()
+	add_undo_operation(create_node_operation)
```
An example of how we could implement the feature flag :)

```suggestion
    if is_keyboards_shortcuts_active:
		var node_position: Vector2 = (graph_view.scroll_offset + graph_view.size * 0.5) / graph_view.zoom
		var create_node_operation: ParleyCreateNodeOperation = ParleyCreateNodeOperation.new(graph_view, type, node_position)
		create_node_operation.do()
		add_undo_operation(create_node_operation)
	else:
		# Previous code
```

JosephStar318 replied:
i see 👍

JosephStar318 replied:
as for the feature flag i believe this will do?

<img width="1163" height="549" alt="image" src="https://github.com/user-attachments/assets/f9c277f0-5bf3-47f1-be15-3ee22a921321" />

JosephStar318 replied:
just saw your comment i will rename the settings to keyboard_shortcuts

jonnydgreen commented on addons/parley/models/dialogue_sequence_ast.gd:
```diff
@@ -144,6 +144,24 @@ func add_new_node(type: Type, position: Vector2 = Vector2.ZERO) -> ParleyNodeAst
 	_emit_dialogue_updated()
 	return ast_node
 
+func copy_node_ast(ast_node: ParleyNodeAst) -> ParleyNodeAst:
```
See my main review above :)

<img width="1215" height="932" alt="Image" src="https://github.com/user-attachments/assets/108436f2-f160-4c4d-a7b0-0181973bfd8a" />

JosephStar318 replied:
i believe this changes will fix the edge creation error after copy?

JosephStar318 replied:
i think it did

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -29,17 +34,18 @@ func _exit_tree() -> void:
 
 
 func clear() -> void:
-	clear_connections()
```
Could we revert or put this removal behind a feature flag? Thanks! :)

JosephStar318 replied:
i don't know why i removed this. might have been get caught up with other clear functions i wrote and delete. let me revert it.

jonnydgreen reviewed (commented):
Hello! Happy March! :) This is soooo good! I've been having a load of fun with it today! We are very close now!

Thanks for implementing the copy/paste functionality, I've just tried it out now! 🚀 On the whole it's working great! I've had a few intermittent issues with it unfortunately and am unable to create edges on copied nodes. Here are my steps to reproduce, lemme know what you think:
 1. Copy node or duplicate node
 2. Try and connect an edge to it

I believe the issue is with the use of `duplicate` and references. Having looked in depth at this, I think we will need to adjust to a more reliable approach as follows (this also gets rid of some of the other issues I was seeing):

<img width="1215" height="932" alt="Image" src="https://github.com/user-attachments/assets/add70a25-6bc0-47c8-bd4c-8919d58e67c0" />
<img width="918" height="236" alt="Image" src="https://github.com/user-attachments/assets/02bd536c-50a7-4837-a5d7-2e9298f29eb2" />


I'm also seeing a few other intermittent issues crop up here and there which are hard to reproduce - (mainly when testing the undo/redo functionality and playing around with edges). Given this and the best kind of testing is using this for real, I think we should put this behind a feature flag in the Parley settings as an opt-in feature (I'm gonna do this with my translations support PR as well). With this in place, this will allow us to release this very soon and once we are fully confident, we can switch to on by default! :)

As I mentioned in my previous comment on this, this can be done in: `parley_settings.gd` and `constants.gd` (if you agree, I recommend: `"parley/editor/keyboard_shortcuts"`) but lemme know if you need a hand or guidance with this

As I understand it, we have the following left to do:
 - Fix any remaining bugs/implementation details
 - Tidy up code, and remove/fix any prints/todos
 - Fix any GDScript warnings/errors
 - Put this behind a feature flag and set to opt-in (add a TODO to switch this to opt-out)
 - Testing - after which, I can do a final bit of review polish
 - Documentation (can be done in a follow-up PR - I'm happy to do this if you'd prefer)

Anything I've missed for what is left to do? :)

I'll aim to spend Thursday/Friday evening of this week doing some further testing this feature and get us ready for merging - really, really well done!! And thank you so much! 🚀

JosephStar318 committed:
Update addons/parley/models/dialogue_sequence_ast.gd

Co-authored-by: Jonny Green <me+github@jonnydgreen.com>

JosephStar318 committed:
Update addons/parley/models/dialogue_sequence_ast.gd

Co-authored-by: Jonny Green <me+github@jonnydgreen.com>

Commit:
- comment cheange on id setter method on node_ast

Commit:
- added feature flag and  made shortcuts an optional feature

- copy shortcut fixes and comment changes

JosephStar318 commented:
I send my changes. Added the keyboard shortcuts opt-in in project settings as you mentioned.

jonnydgreen commented:
> I send my changes. Added the keyboard shortcuts opt-in in project settings as you mentioned.

Fantastic, thank you so much! I will take a look this week :)

jonnydgreen commented:
Hello! Apologies for the delay in reviewing this - I had/have to focus on some other things outside of OSS but I should be all good to test, review, and release this in a couple of weeks! :)

jonnydgreen commented on addons/parley/models/dialogue_sequence_ast.gd:
```diff
@@ -241,12 +265,12 @@ func remove_node(node_id: String) -> void:
 func find_node_by_id(id: String) -> ParleyNodeAst:
 	var filtered_nodes: Array = nodes.filter(func(node: ParleyNodeAst) -> bool: return str(node.id) == str(id))
 	if filtered_nodes.size() != 1:
+		print("filtered node count",filtered_nodes.size())
```
Can we remove this log? :)

jonnydgreen commented on addons/parley/models/dialogue_sequence_ast.gd:
```diff
@@ -635,7 +659,10 @@ static func is_dialogue_options(p_nodes: Array[ParleyNodeAst]) -> bool:
+	if is_instance_of(raw_position, TYPE_VECTOR2):
+		return raw_position
+	elif not is_instance_of(raw_position, TYPE_STRING):
+		print(raw_position)
```
Can we remove this log please? :)

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -312,6 +319,7 @@ func _create_group_node(ast_node: ParleyGroupNodeAst, _should_regenerate: bool =
 	# EXPERIMENTAL: see how feedback goes. This is certainly a candidate to be put into settings
 	ParleyUtils.signals.safe_connect(node.dragged, func(_from: Vector2, _to: Vector2) -> void:
 		var _nodes: Array[ParleyGraphNode] = _update_nodes_covered_by_group_node(node, ast_node)
+		print("dragged")
```
Can we remove this log please? :)

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -351,4 +359,290 @@ func set_selected_by_id(id: String, _goto: bool = true) -> void:
+		unselected_node.call_deferred("set_selected", false)
+		return
+	selected_node_ids.append(node_id)
+	# print("selected callback ", node_id, " ", selected_node_ids.size())
```
Can we remove this comment please? :)

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -351,4 +359,290 @@ func set_selected_by_id(id: String, _goto: bool = true) -> void:
+func _on_node_deselected(node: Node) -> void:
+	if not ParleySettings.get_setting(Constants.EDITOR_IS_KEYBOARD_SHORTCUTS_ACTIVE, false): 
+		return
+	# print("deselected callback ", selected_node_ids.size())
```
Can we remove this comment please? :)

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -351,4 +359,290 @@ func set_selected_by_id(id: String, _goto: bool = true) -> void:
+
+
+func _try_select_edge(pointer_pos: Vector2, is_control_pressed: bool) -> bool:
+	# print("pointer_pos ", pointer_pos," control pressed: ", is_control_pressed)
```
Can we remove this comment please? :)

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -351,4 +359,290 @@ func set_selected_by_id(id: String, _goto: bool = true) -> void:
+		var p2: Vector2 = controls[2]
+		var p3: Vector2 = controls[3]
+		var distance: float = get_distance_to_bezier(pointer_pos, p0, p1, p2, p3) * zoom
+		# print("From: ",from_node ," To", to_node," Comparing x: ", from_pos.x, " ", pointer_pos.x, " ", to_pos.x , " distance: ", distance)
```
Can we remove this comment please? :)

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -351,4 +359,290 @@ func set_selected_by_id(id: String, _goto: bool = true) -> void:
+	for node_id : String in selected_node_ids:
+		var node : ParleyGraphNode = find_node_by_id(node_id)
+		var node_rect: Rect2 = Rect2(node.position, node.size / zoom)
+		print("pos",pos, "rect: ",node_rect)
```
Can we remove this log please? :)

jonnydgreen commented on addons/parley/main_panel.gd:
```diff
@@ -599,6 +619,7 @@ func update_edge(edge: ParleyEdgeAst) -> void:
 
 # TODO: add to docs
 func delete_node_by_id(id: String) -> void:
+	print(id)
```
Can we remove this log please? :)

jonnydgreen commented on addons/parley/graph_operations/parley_duplicate_operation.gd:
```diff
@@ -0,0 +1,66 @@
+			created_node_list.append(new_ast_node.id)
+			node_names.append(graph_view.get_ast_node_name(new_ast_node))
+
+		# var selected_ast : ParleyNodeAst = graph_view.ast.find_node_by_id(node_id)
+		# var ast_node: ParleyNodeAst = graph_view.ast.add_new_node(selected_ast.type, selected_ast.position + Vector2.DOWN * 200 + Vector2.RIGHT * 200 )
+		# if ast_node:
+		# 	node_id = ast_node.id
+		# 	var connections : Array[ParleyGraphEdge] = ParleyGraphUtils.get_connections_for_node(graph_view, node_id)
+		# 	created_node_datas.append(NodeData.new(node_id, ast_node, connections))
+		# 	created_node_list.append(node_id)
+		# 	node_names.append(selected_ast.resource_name)
```
Can we remove these comments please? :)

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -351,4 +359,290 @@ func set_selected_by_id(id: String, _goto: bool = true) -> void:
+		var pointer_pos: Vector2 = (mouse_event.position + scroll_offset) / zoom
+		var is_control_pressed: bool = mouse_event.is_command_or_control_pressed()
+
+		# if _node_exist_at_position(mouse_event.position): 
+		# 	_unselect_edges()
+		# 	return
```
Can we remove these comments please? :)

jonnydgreen commented on addons/parley/models/node_ast.gd:
```diff
@@ -35,11 +35,11 @@ func _init(_id: String = "", _position: Vector2 = Vector2.ZERO) -> void:
+	if id.begins_with(id_prefix):
+		id = new_id
+	elif not id or id == id_prefix or not id.begins_with(id_prefix):
 		id = new_id if new_id.begins_with(id_prefix) else "%s%s" % [id_prefix, new_id]
```
Could we add `#endregion` back please? :)

```suggestion
		id = new_id if new_id.begins_with(id_prefix) else "%s%s" % [id_prefix, new_id]
#endregion
```

jonnydgreen commented on addons/parley/views/graph_connection.gd.uid:
```diff
@@ -0,0 +1 @@
+uid://bb4mruigkoqs5
```
Is this file still needed?

JosephStar318 replied:
nope its a resiude.

jonnydgreen commented on addons/parley/views/parley_graph_view.gd:
```diff
@@ -19,8 +20,13 @@ const end_node_scene: PackedScene = preload("../components/end/end_node.tscn")
+
 # Called when the node enters the scene tree for the first time.
-func _ready() -> void:
+func _ready() -> void:	
```
```suggestion
func _ready() -> void:
```

jonnydgreen commented on addons/parley/main_panel.gd:
```diff
@@ -50,11 +49,13 @@ var parley_manager: ParleyManager
 var selected_node_id: Variant
 var selected_node_ast: ParleyNodeAst: set = _set_selected_node_ast
-
+var copied_node_ids: Array[String]
```
Just to be on the safe side:

```suggestion
var copied_node_ids: Array[String] = []
```

jonnydgreen reviewed (approved):
Fantastic! Looking over the bezier stuff again was great! 🔥 I've left a few comments to tidy up the code a bit and provide a bit of extra safety, but I think we are there!

The only thing remaining is some documentation but that can be a follow-up PR at some later time - wext stop is merge (squash) and release I think! Well done and thank you!

JosephStar318 committed:
Update addons/parley/models/node_ast.gd

#endregion added back.

Co-authored-by: Jonny Green <me+github@jonnydgreen.com>

JosephStar318 committed:
Update addons/parley/views/parley_graph_view.gd

Co-authored-by: Jonny Green <me+github@jonnydgreen.com>

Commit:
-removed some logs and refresh edges will only be active if shortcuts enabled

Commit:
Merge branch 'keyboard-shortcuts' of https://github.com/JosephStar318/parley into keyboard-shortcuts

JosephStar318 committed:
Update addons/parley/main_panel.gd

Co-authored-by: Jonny Green <me+github@jonnydgreen.com>
